### PR TITLE
Update to Python 3.8 minimum.

### PR DESCRIPTION
**Pull request by bjrichardwebster**

### CODE REVIEW

File: docs/release_notes/pending_release.rst [hunk 1]
```diff
@@ -4,10 +4,15 @@ Pending Release Notes
...
 
+Python
+
+* New minimum supported python changed to `python = "^3.8"`.
+
 CI
 
 * Updated CI unittests workflow to include codecov reporting.
   Reduced CodeCov report submission by skipping this step on scheduled runs.
+* Updated CI unittests to reflect new minimum support `python = "^3.8"`.
```
Comment (Purg) [inline docs/release_notes/pending_release.rst:15]: Minor: sphinx/RST is going to want a newline between bullets.

File: pyproject.toml [hunk 2]
```diff
@@ -18,39 +18,40 @@ classifiers = [
...
 ]
 
 [tool.poetry.dependencies]
-python = "^3.6"
-numpy = "^1.19.5"
-pillow = "^8.3.2"
-smqtk-core = ">=0.18.0"
+python = "^3.8"
+pillow = ">=9.2.0"
+numpy = ">=1.22"
```
Comment (Purg) [inline pyproject.toml:30]: Are these updated version lower-bounds required? I.e. could we instead have `>=` inequalities with the previous versions?

File: pyproject.toml [hunk 3]
```diff
@@ -18,39 +18,40 @@ classifiers = [
...
+livereload = ">=2.6.3"
 # Testing
 coverage = "^5.5"
-mock = "^4.0.3"
-pytest = "^6.2.2"
-pytest-cov = "^2.11.1"
+mock = ">=4.0.3"
+pytest = ">=7.2.0"
+pytest-cov = ">=2.11.1"
+requests = ">=2.31.0"
```
Comment (Purg) [inline pyproject.toml:52]: Looks like a new dependency. Is this required due to some ripple effect?

**Review by Purg — COMMENTED** [review]
Were you thinking you were going to address the new mypy typechecking issues with this branch?

File: pyproject.toml (hunk 2 above)
Comment (bjrichardwebster) [reply pyproject.toml:30]: Pillow:
There were numerous dependabot alerts, this version is the lowest to address them all.

Numpy:
Dependabot alert.

Neither are transitory packages.

File: pyproject.toml (hunk 3 above)
Comment (bjrichardwebster) [reply pyproject.toml:52]: The change from ^ to >= is to allow versions for python3.11, the requests import is a mistake I'll remove.

**Comment by bjrichardwebster** [comment]
When addressing the mypy type checking errors I came across three major themes:

1) numpy added a couple different explicit typing features in 1.20, 1.21, and 1.22. This caused a few cascading type checking issues that I've addressed by Union between the numpy typing definition with the old Sequence definition which handles non-numpy inputs

2) A couple of test files were attempting to override an abstract function with Optional parameter with a non-Optional function. This is no longer allowed because mypy has changed its default to no_implicit_optional=True. See [incompatible-overrides](https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides).

3) ImageReader.load_as_matrix has an Optional return while numpy.testing.assert_allclose parameters do not allow Optional. Since all mypy errors connected to this difference were in the unit tests, I added the `type: ignore` to those lines. In the unit tests we know None will never be returned in those situations and we want to keep the unit test focused, so adding an extra explicit assertion for None (instead of type:ignore) would add an assertion outside the scope of the particular unit test.

**Review by Purg — APPROVED** [review]

**Pull request merged**